### PR TITLE
[Refactor] Proxy/Request namespace and context

### PR DESCRIPTION
**Pull request by alagroy-42**
#1936 with the new files already moved.

### CODE REVIEW

File: lizmap/modules/lizmap/lib/Request/WFSRequest.php [hunk 1]
```diff
@@ -12,6 +12,9 @@
 
 namespace Lizmap\Request;
 
+/**
+ * https://en.wikipedia.org/wiki/Web_Feature_Service.
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/WFSRequest.php:16]: the url should be behind `@see`

```
* @see https://en.wikipedia.org/wiki/Web_Feature_Service.
```

File: lizmap/modules/lizmap/lib/Request/WFSRequest.php [hunk 2]
```diff
@@ -288,6 +300,10 @@ public function getfeature()
         return $this->getfeatureQgis();
     }
 
+    /**
+     * https://en.wikipedia.org/wiki/Web_Feature_Service#Static_Interfaces
+     * Queries Qgis Server for getFeature.
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/WFSRequest.php:305]: decription should be above the `@see` (that you should add for the url).

Fix this for all comments you added

File: lizmap/modules/lizmap/lib/Request/Proxy.php [hunk 3]
```diff
@@ -21,6 +21,94 @@ class Proxy
      */
     protected static $_profiles = array();
 
+    protected static $services = null;
+
+    protected static $appContext = null;
+
+    public static function setServices($services = null)
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/Proxy.php:28]: The function name is not related on how the method is used, as it is used to get services.
I think we should have a `getServices()`, which return the current `self::$services` if not null,else `\lizmap::getServices()`. And we should have a `setServices($services)` to set  `self::$services`.

File: lizmap/modules/lizmap/lib/Request/Proxy.php [hunk 4]
```diff
@@ -21,6 +21,94 @@ class Proxy
...
+                self::$services = $services;
+            } else {
+                self::$services = \lizmap::getServices();
+            }
+        }
+
+        return self::$services;
+    }
+
+    public static function setAppContext($appContext = null)
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/Proxy.php:41]: same as setServices: we should have a `getAppContext()` and a `setAppContext($appContext)`

File: lizmap/modules/lizmap/lib/Request/Proxy.php [hunk 5]
```diff
@@ -725,19 +528,24 @@ protected static function declareRedisProfile($ser, $cacheName, $repository, $pr
         }
 
         // Create the virtual cache profile
-        jProfiles::createVirtualProfile('jcache', $cacheName, $cacheParams);
+        self::setAppContext()->createVirtualProfile('jcache', $cacheName, $cacheParams);
     }
 
     /**
      * @param mixed $repository
+     * @param mixed $lrep
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/Proxy.php:536]: the previous @param should be removed. And the type should be indicated.

File: lizmap/modules/lizmap/lib/Request/WMSRequest.php [hunk 6]
```diff
@@ -847,4 +866,292 @@ protected function gfiGmlToHtml($gmldata, $configLayer)
...
+
+            // Keep original param value
+            $originalParams = array('width' => $params['width'], 'height' => $params['height']);
+            // Replace width and height before requesting the image from qgis
+            $params['width'] = $metatileSizeX * $params['width'] + 2 * $metatileBuffer;
+            $params['height'] = $metatileSizeY * $params['height'] + 2 * $metatileBuffer;
+        }
+
+        // Build params before send the request to Qgis
+        $builtParams = http_build_query($params);
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/WMSRequest.php:1061]: Probably we could use Proxy::constructUrl() here ?

**Review by laurentj — CHANGES_REQUESTED** [review]

File: lizmap/modules/lizmap/lib/Request/WMSRequest.php (hunk 6 above)
Comment (alagroy-42) [reply lizmap/modules/lizmap/lib/Request/WMSRequest.php:1061]: Done in next PR.

Comment (laurentj) [reply lizmap/modules/lizmap/lib/Request/WMSRequest.php:1061]: okay

File: lizmap/modules/lizmap/lib/Request/Proxy.php [hunk 7]
```diff
@@ -21,6 +21,96 @@ class Proxy
      */
     protected static $_profiles = array();
 
+    protected static $services = null;
+
+    protected static $appContext = null;
+
+    public static function setServices($services)
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/Proxy.php:28]: Type of parameters is missing (in an `@param` or directly into the code).
Please document parameters and return values for all new methods.

**Review by laurentj — CHANGES_REQUESTED** [review]
still have some issues on documentation. It will be ok after that

File: lizmap/modules/lizmap/lib/Request/Proxy.php (hunk 7 above)
Comment (laurentj) [reply lizmap/modules/lizmap/lib/Request/Proxy.php:34]: the syntax is `@param <type> <parameter name>`

`@param \lizmapServices $services`. No two `@param` for the same parameter.

**Pull request merged**